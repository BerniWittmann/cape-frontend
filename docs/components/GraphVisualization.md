# GraphVisualization

A Component which visualizes the connection between Processes, Context Situations and Context Factors

