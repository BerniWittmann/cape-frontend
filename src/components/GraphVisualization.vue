<template>
  <div id="graph">
    <svg>
      <g></g>
    </svg>
  </div>
</template>

<script>
import * as d3 from 'd3'
import dagreD3 from 'dagre-d3'
import { mapGetters } from 'vuex'

/*
 * @vuese
 * @group Components
 *
 * A Component which visualizes the connection between Processes, Context Situations and Context Factors
 */
export default {
  name: 'GraphVisualization',
  computed: {
    ...mapGetters({
      'processes': 'process/processGraphNodes',
      'contextSituations': 'contextSituation/contextSituationGraphNodes',
      'contextFactors': 'contextFactor/contextFactorGraphNodes',
      'edges': 'graphEdges'
    }),

    nodes() {
      return [...this.processes, ...this.contextSituations, ...this.contextFactors]
    }
  },

  methods: {
    setNode(entity) {
      this.g.setNode(entity.id, { label: entity.name, class: entity.type, id: entity.id, route: entity.route })
      this.g.setParent(entity.id, `group_${entity.type}`)
    },
    setEdge(edge) {
      this.g.setEdge(edge.start, edge.end, { arrowhead: 'undirected', curve: d3.curveMonotoneX, class: edge.type })
    },
    centerGraph() {
      const xCenterOffset = (this.$el.clientWidth - this.g.graph().width) / 2
      this.svgGroup.attr('transform', 'translate(' + xCenterOffset + ', 20)')
      this.svg.attr('height', this.g.graph().height + 40)
    },
    resizeHandler() {
      this.centerGraph()
    }
  },

  beforeDestroy() {
    window.removeEventListener('resize', this.resizeHandler)
  },

  mounted() {
    window.addEventListener('resize', this.resizeHandler)
    this.g = new dagreD3.graphlib.Graph({ compound: true })
    this.g.setGraph({
      nodesep: 30,
      ranksep: 100,
      rankdir: 'LR',
      align: 'UL',
      marginx: 20,
      marginy: 20
    })
    this.g.setDefaultEdgeLabel(() => ({}))

    this.g.setNode('group_process', { label: 'Processes', clusterLabelPos: 'top' })
    this.g.setNode('group_situation', { label: 'Situations', clusterLabelPos: 'top' })
    this.g.setNode('group_factor', { label: 'Factors', clusterLabelPos: 'top' })

    this.nodes.forEach((entity) => {
      this.setNode(entity)
    })
    this.g.nodes().forEach((v) => {
      const node = this.g.node(v)
      // Round the corners of the nodes
      node.rx = node.ry = 5
    })

    this.edges.forEach((edge) => {
      this.setEdge(edge)
    })

    // eslint-disable-next-line new-cap
    const render = new dagreD3.render()

    this.svg = d3.select('#graph svg')
    this.svgGroup = this.svg.select('g')

    render(this.svgGroup, this.g)

    this.centerGraph()

    this.labels = this.svg.selectAll('g.nodes > *')
    this.labels.on('mouseover', (id) => {
      const node = this.g.node(id)
      node.elem.classList.add('active')

      const paths = this.g.nodeEdges(id)
      paths.forEach((p) => {
        const edge = this.g.edge(p)
        edge.elem.classList.add('active')
      })

      const edges = this.g.edges()
      edges.forEach((e) => {
        const edge = this.g.edge(e)
        if (!edge.elem.classList.contains('active')) {
          edge.elem.classList.add('inactive')
        }
      })
    })
    this.labels.on('mouseout', (id) => {
      const node = this.g.node(id)
      node.elem.classList.remove('active')

      const edges = this.g.edges()
      edges.forEach((e) => {
        const edge = this.g.edge(e)
        edge.elem.classList.remove('active')
        edge.elem.classList.remove('inactive')
      })
    })
    this.labels.on('click', (id) => {
      this.$router.push(this.g.node(id).route)
    })
    this.svg.selectAll('.edgePath path').attr('marker-end', '').attr('stroke-linecap', 'round')
  }
}

</script>

<style lang="scss">
$colors: (situation: #03A6FF, process: #1EC0FF, factor: #0080FF);

#graph {
  svg {
    width: 100%;

    .node {
      @each $name, $color in $colors {
        &.#{$name} {
          rect {
            fill: rgba($color, 0.1);
            stroke: rgba($color, 0.2);
          }

          text {
            fill: rgba($color, 1);
          }
        }
      }
    }
  }

  text {
    font-weight: 300;
    font-size: 14px;
  }

  .node {
    cursor: pointer;

    &.active {
      rect {
        transform: scale(1.1);
      }
    }

    rect {
      transition: all 200ms ease-in-out;
      fill: #fff;
    }
  }

  .edgePath {
    path {
      stroke: #8e8e8e;
      stroke-width: 5px;
      transition: all 200ms ease-in-out;
    }

    &.active {
      path {
        opacity: 1;
      }
    }

    &.inactive {
      path {
        opacity: 0.1;
      }
    }

    &.process_situation {

    }

    &.situation_factor {

    }
  }

  .clusters rect {
    fill: none;
    stroke: none;
  }

}

</style>
