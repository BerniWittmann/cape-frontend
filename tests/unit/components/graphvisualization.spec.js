import { shallowMount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'
import * as d3 from 'd3'
import dagreD3 from 'dagre-d3'

import GraphVisualization from '@/components/GraphVisualization.vue'

describe('Components', () => {
  describe('GraphVisualization', () => {
    let store
    let cmp
    let graph
    let router

    beforeEach(() => {
      store = {
        getters: {
          'process/processGraphNodes': [{
            id: 'p1',
            name: 'Process 1',
            type: 'process',
            route: {
              name: 'process.preview',
              params: {
                processID: 'p1'
              }
            }
          }, {
            id: 'p2',
            name: 'Process 2',
            type: 'process',
            route: {
              name: 'process.preview',
              params: {
                processID: 'p2'
              }
            }
          }, {
            id: 'p3',
            name: 'Process 3',
            type: 'process',
            route: {
              name: 'process.preview',
              params: {
                processID: 'p3'
              }
            }
          }],
          'contextFactor/contextFactorGraphNodes': [{
            id: 'cF1',
            name: 'Factor 1',
            type: 'factor',
            route: {
              name: 'context_factor.edit',
              params: {
                processID: 'cF1'
              }
            }
          }, {
            id: 'cF2',
            name: 'Factor 2',
            type: 'factor',
            route: {
              name: 'context_factor.edit',
              params: {
                processID: 'cF1'
              }
            }
          }, {
            id: 'cF3',
            name: 'Factor 3',
            type: 'factor',
            route: {
              name: 'context_factor.edit',
              params: {
                processID: 'cF3'
              }
            }
          }, {
            id: 'cF4',
            name: 'Factor 4',
            type: 'factor',
            route: {
              name: 'context_factor.edit',
              params: {
                processID: 'cF4'
              }
            }
          }],
          'contextSituation/contextSituationGraphNodes': [{
            id: 'cS1',
            name: 'Situation 1',
            type: 'situation',
            route: {
              name: 'context_situation.single',
              params: {
                processID: 'cS1'
              }
            }
          }, {
            id: 'cS2',
            name: 'Situation 2',
            type: 'situation',
            route: {
              name: 'context_situation.single',
              params: {
                processID: 'cS2'
              }
            }
          }],
          graphEdges: [{
            start: 'p1',
            end: 'cS1',
            type: 'process_situation'
          }, {
            start: 'p1',
            end: 'cS2',
            type: 'process_situation'
          }, {
            start: 'p3',
            end: 'cS2',
            type: 'process_situation'
          }, {
            start: 'cS1',
            end: 'cF1',
            type: 'situation_factor'
          }, {
            start: 'cS2',
            end: 'cF3',
            type: 'situation_factor'
          }, {
            start: 'cS1',
            end: 'cF3',
            type: 'situation_factor'
          }, {
            start: 'cS2',
            end: 'cF2',
            type: 'situation_factor'
          }]
        }
      }
      router = {
        push: jest.fn()
      }
      window.addEventListener = jest.fn()
      window.removeEventListener = jest.fn()
      jest.clearAllMocks()
      render()
    })

    function render() {
      cmp = shallowMount(GraphVisualization, {
        i18n,
        mocks: {
          $store: store,
          $router: router
        }
      })
      graph = cmp.vm.g
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('renders an svg', () => {
      const graph = cmp.find('#graph')
      expect(graph.exists()).toBeTruthy()
      const svg = graph.find('svg')
      expect(svg.exists()).toBeTruthy()
    })

    it('setups the graph', () => {
      expect(graph.options).toEqual({ compound: true })
      expect(graph.graphOptions).toEqual(expect.any(Object))

      expect(graph.defaultEdgeLabel).toEqual(expect.any(Function))
      expect(graph.defaultEdgeLabel()).toEqual({})
    })

    it('renders', () => {
      expect(dagreD3.renderFunction.mock.calls[0][0].isD3).toBeTruthy()
      expect(dagreD3.renderFunction.mock.calls[0][1].isGraph).toBeTruthy()
    })

    it('renders all nodes', () => {
      expect(Object.keys(graph._nodes).length).toEqual(12)
    })

    describe('renders the processes', () => {
      it('renders a process group', () => {
        const group = graph._nodes['group_process']
        expect(group).toEqual({
          'clusterLabelPos': 'top',
          'label': 'Processes',
          'rx': 5,
          'ry': 5,
          elem: expect.any(Object)
        })
      })

      it('renders all processes', () => {
        store.getters['process/processGraphNodes'].forEach(p => {
          const process = graph._nodes[p.id]
          expect(process).toEqual({
            id: p.id,
            label: p.name,
            class: 'process',
            route: p.route,
            rx: 5,
            ry: 5,
            parent: 'group_process',
            elem: expect.any(Object)
          })
        })
      })
    })

    describe('renders the factors', () => {
      it('renders a factors group', () => {
        const group = graph._nodes['group_factor']
        expect(group).toEqual({
          'clusterLabelPos': 'top',
          'label': 'Factors',
          'rx': 5,
          'ry': 5,
          elem: expect.any(Object)
        })
      })

      it('renders all factors', () => {
        store.getters['contextFactor/contextFactorGraphNodes'].forEach(f => {
          const factor = graph._nodes[f.id]
          expect(factor).toEqual({
            id: f.id,
            label: f.name,
            class: 'factor',
            route: f.route,
            rx: 5,
            ry: 5,
            parent: 'group_factor',
            elem: expect.any(Object)
          })
        })
      })
    })

    describe('renders the situations', () => {
      it('renders a situations group', () => {
        const group = graph._nodes['group_situation']
        expect(group).toEqual({
          'clusterLabelPos': 'top',
          'label': 'Situations',
          'rx': 5,
          'ry': 5,
          elem: expect.any(Object)
        })
      })

      it('renders all situations', () => {
        store.getters['contextSituation/contextSituationGraphNodes'].forEach(s => {
          const situation = graph._nodes[s.id]
          expect(situation).toEqual({
            id: s.id,
            label: s.name,
            class: 'situation',
            route: s.route,
            rx: 5,
            ry: 5,
            parent: 'group_situation',
            elem: expect.any(Object)
          })
        })
      })
    })

    describe('renders the edges', () => {
      it('renders all edges', () => {
        expect(Object.keys(graph._edges).length).toEqual(7)
      })
    })

    describe('listens to node mouse over', () => {
      let handler
      beforeEach(() => {
        handler = cmp.vm.labels._listeners['mouseover']
      })

      it('listens for the mouseover event', () => {
        expect(handler).toEqual(expect.any(Function))
      })
      it('adds the active class to the current node', () => {
        handler('p1')

        expect(graph.node('p1').elem.classList.contains('active')).toBeTruthy()
      })

      it('adds the active classes to all adjoining edges', () => {
        handler('p1')

        graph.nodeEdges().forEach(e => {
          e = graph.edge(e)
          expect(e.elem.classList.contains('active')).toBeTruthy()
          expect(e.elem.classList.contains('inactive')).toBeFalsy()
        })
      })

      it('adds the inactive classes to all non adjoining edges', () => {
        handler('p1')

        const nodeEdges = graph.nodeEdges('p1')
        graph.edges().forEach(e => {
          const ed = graph.edge(e)
          if (!nodeEdges.includes(e)) {
            expect(ed.elem.classList.contains('inactive')).toBeTruthy()
            expect(ed.elem.classList.contains('active')).toBeFalsy()
          }
        })
      })
    })

    describe('unsets highlights on mouse leave', () => {
      let handler
      beforeEach(() => {
        handler = cmp.vm.labels._listeners['mouseout']
        graph.node('p2').elem.classList.add('active')
        graph.edges().forEach(e => {
          e = graph.edge(e)
          const state = Math.random() > 0.5 ? 'active' : 'inactive'
          e.elem.classList.add(state)
        })
      })
      it('listens for the mouseout event', () => {
        expect(handler).toEqual(expect.any(Function))
      })
      it('removes the active class from the active node', () => {
        handler('p2')

        expect(graph.node('p2').elem.classList.contains('active')).toBeFalsy()
      })

      it('removes the active and inactive classes from all edges', () => {
        handler('p2')

        graph.edges().forEach(e => {
          e = graph.edge(e)
          expect(e.elem.classList.contains('active')).toBeFalsy()
          expect(e.elem.classList.contains('inactive')).toBeFalsy()
        })
      })
    })

    it('navigates to route on node click', () => {
      const clickHandler = cmp.vm.labels._listeners['click']
      expect(clickHandler).toEqual(expect.any(Function))

      const nodes = [
        ...store.getters['process/processGraphNodes'],
        ...store.getters['contextFactor/contextFactorGraphNodes'],
        ...store.getters['contextSituation/contextSituationGraphNodes']
      ]
      nodes.forEach(node => {
        clickHandler(node.id)
        expect(router.push).toHaveBeenCalledWith(node.route)
      })
    })

    describe('it centers the graph on resize', () => {
      let handler
      beforeEach(() => {
        handler = window.addEventListener.mock.calls[0][1]
      })
      it('centers the graph on render', () => {
        expect(d3.default._attrs).toEqual({
          height: 940,
          transform: 'translate(-512, 20)',
          'marker-end': '',
          'stroke-linecap': 'round'
        })
      })

      it('has a lister on the resize event', () => {
        expect(handler).toEqual(expect.any(Function))
        expect(window.addEventListener.mock.calls[0]).toEqual(['resize', cmp.vm.resizeHandler])
      })

      it('centers the graph on resize', () => {
        d3.default._attrs = {}

        expect(d3.default._attrs).toEqual({})

        handler()

        expect(d3.default._attrs).toEqual({
          height: 940,
          transform: 'translate(-512, 20)'
        })
      })

      it('removes the resize listener on destroy', () => {
        expect(window.removeEventListener).not.toHaveBeenCalled()

        cmp.destroy()

        expect(window.removeEventListener).toHaveBeenCalledWith('resize', cmp.vm.resizeHandler)
      })
    })
  })
})
