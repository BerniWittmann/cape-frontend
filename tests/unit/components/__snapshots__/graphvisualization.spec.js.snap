// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components GraphVisualization renders 1`] = `
<div id="graph"><svg>
    <g></g>
  </svg></div>
`;
